### PR TITLE
Fix distinct unsoundness

### PR DESCRIPTION
**Pull request by Halbaroth**
This PR quickly fixes the clash between the SMT-LIB specification of `distinct` and the implementation of this expression in AE.

Basically, we expand the expression `distinct x_1 ... x_n` into a conjunction of binary disequalities.

Notice that the model outputted for `uf2.models.smt2` changes because I don't preserve the old order used in `distinct`.

Fix issue #889 for the `v2.5.x` branch.

### CODE REVIEW

**Comment by bclement-ocp** [comment]
> This fix doesn't change the smart constructor `mk_distinct`. We shouldn't change the API on the branch v2.5.x and I add a comment to advertise the user that this smart constructor doesn't behave as expected.

I disagree on this point. `mk_distinct` is supposed to build a `distinct` term; even if the API is undocumented, the intent (as expressed by the name `mk_distinct`) is clear to any reasonable user. `mk_distinct` not building a an "all-distinct" term *is* the bug that should be fixed.

(One other way to think about it: if we have a function `mk_plus` that actually created a multiplication, we would fix it in the function itself, we would not add workarounds)

**Comment by Halbaroth** [comment]
Besides, I checked that the constructor `TANeq` of the legacy parser is never used with a list of at least three elements.

File: src/lib/structures/expr.ml [hunk 1]
```diff
@@ -1210,10 +1210,20 @@ let mk_nary_eq l =
...
 
-let mk_distinct ~iff tl =
-  match tl with
-  | [a; b] -> neg (mk_eq ~iff a b)
-  | _ -> neg (mk_nary_eq tl)
+let mk_distinct ~iff args =
+  (* This fix makes sure that the smart constructor agrees with
+     the SMT-LIB specification when used with at least 3 arguments.
+     To prevent a soundness bug, we translate the expected expression into a
+     conjonction of binary disequations. *)
```
Comment (bclement-ocp) [inline src/lib/structures/expr.ml]: Note: this is not specifically related to the SMT-LIB specification but to the usual definition of `distinct`.

File: src/lib/structures/expr.ml [hunk 2]
```diff
@@ -1210,10 +1210,20 @@ let mk_nary_eq l =
...
+  let args = Array.of_list args in
+  let acc = ref vrai in
+  for i = 0 to Array.length args - 1 do
+    for j = i + 1 to Array.length args - 1 do
+      acc :=
+        mk_and (neg (mk_eq ~iff args.(i) args.(j))) !acc false
+    done;
+  done;
+  !acc
```
Comment (bclement-ocp) [inline src/lib/structures/expr.ml:1227]: Nit: it doesn't really matter here as this is a hotfix but conversion to an array is not needed:

```ocaml
let distinct_from x acc ys =
  List.fold_left (fun acc y -> mk_and (neg (mk_eq ~iff x y)) acc) acc ys
in let rec distinct_aux acc = function
  | [] -> acc
  | x :: ys -> distinct_aux (distinct_from x acc ys) ys
in distinct_aux acc args
```

File: src/lib/frontend/cnf.ml [hunk 3]
```diff
@@ -239,9 +239,13 @@ and make_form up_qv name_base ~toplevel f loc ~decl_kind : E.t =
...
           if negated then E.neg res else res
 
-        | TAneq lt | TAdistinct lt ->
+        | TAneq lt ->
           let lt = List.map (make_term up_qv name_base) lt in
           E.mk_distinct ~iff:true lt
+
+        | TAdistinct lt ->
+          E.mk_distinct ~iff:true (List.map (make_term up_qv name_base) lt)
+
```
Comment (bclement-ocp) [inline src/lib/frontend/cnf.ml]: Can we avoid these changes? They don't seem useful, and keeping the patch as tiny as possible makes it easier to cherry-pick and move around if needed.

File: src/lib/frontend/d_cnf.ml [hunk 4]
```diff
@@ -1337,7 +1337,7 @@ let rec mk_expr
             end
 
           | B.Distinct, _ ->
-            E.mk_distinct ~iff:true (List.map (fun t -> aux_mk_expr t) args)
+            E.mk_distinct ~iff:true (List.map aux_mk_expr args)
```
Comment (bclement-ocp) [inline src/lib/frontend/d_cnf.ml]: Same as above, can we undo this? There shouldn't be this sort of unrelated cleanup in a hotfix.

File: src/lib/structures/expr.ml (hunk 1 above)
Comment (bclement-ocp) [reply src/lib/structures/expr.ml]: Also the comment should point to #889 which will track the fix in the `next` branch

**Review by bclement-ocp — CHANGES_REQUESTED** [review]
> even if the API is undocumented

It actually is: https://ocamlpro.github.io/alt-ergo/Input_file_formats/Native/02_types/02_01_builtin.html#comparisons

Please remove the cleanups that have no place in a hotfix and it should be good.

File: src/lib/structures/expr.ml (hunk 1 above)
Comment (Halbaroth) [reply src/lib/structures/expr.ml]: Actually I didn't plan to apply this patch on next but I can.

Comment (bclement-ocp) [reply src/lib/structures/expr.ml]: Sorry, I maybe was not clear. #889 is not about this patch, #889 is about the issue with `distinct`. We don't need to apply the patch on `next`. We will fix the issue with `distinct` on `next` in a different way, which is fine, but we should still refer to it here.

File: src/lib/structures/expr.ml (hunk 2 above)
Comment (Halbaroth) [reply src/lib/structures/expr.ml:1227]: I know but I think the imperative version is the easier to read and a bit less error-prone. It's a matter of taste.

File: src/lib/structures/expr.ml (hunk 1 above)
Comment (Halbaroth) [reply src/lib/structures/expr.ml]: Ok :)

File: src/lib/frontend/cnf.ml (hunk 3 above)
Comment (Halbaroth) [reply src/lib/frontend/cnf.ml]: Sure. Besides, we should use a different constructor for TAneq? It's not documented and I'm not sure what is the expected semantic for this constructor.

File: src/lib/structures/expr.ml (hunk 2 above)
Comment (bclement-ocp) [reply src/lib/structures/expr.ml:1227]: It is not only a matter of taste (I also find the imperative version easier to read in this specific case); the imperative version is slower and uses more memory because of the call to `Array.of_list`. It is entirely negligible here (hence why it's a nit), just pointing that out.

File: src/lib/frontend/cnf.ml (hunk 3 above)
Comment (bclement-ocp) [reply src/lib/frontend/cnf.ml]: `TAneq [x; y]` is the constructor used for `x <> y`. "neq" is a fairly common shorthand for "**n**ot **eq**ual".

File: src/lib/frontend/cnf.ml [hunk 5]
```diff
@@ -242,6 +242,7 @@ and make_form up_qv name_base ~toplevel f loc ~decl_kind : E.t =
         | TAneq lt | TAdistinct lt ->
           let lt = List.map (make_term up_qv name_base) lt in
           E.mk_distinct ~iff:true lt
+
```
Comment (bclement-ocp) [inline src/lib/frontend/cnf.ml]: Can you undo this whitespace change? Thanks.

**Review by bclement-ocp — CHANGES_REQUESTED** [review]

File: src/lib/structures/expr.ml (hunk 2 above)
Comment (Halbaroth) [reply src/lib/structures/expr.ml:1227]: Sure ;)

**Review by bclement-ocp — APPROVED** [review]

**Review by Stevendeo — COMMENTED** [review]
LGTM!

EDIT: I still had the notification to review the PR after it was merged

**Comment by Halbaroth** [comment]
:D
